Deduplicate pages between the "Popular pages" and "Trending pages" template

icyflame opened the pull request:
- Takes the list of popular pages that we have updated to the template
- Removes any element in `Trending pages` that also occurs in popular pages
- Updates the template to show the required number of trending pages

Fix #30

icyflame committed:
parse all results, display only the required length

icyflame committed:
deduplicate the trending list, maintaining order

deduping t and s using list(set(t) - set(s)) doesn't maintain the order. here
order is important to us (since it is sorted by the metrics we give GA)

icyflame committed:
update the template pages, remove debug logs

amrav commented on jobs/pywikibot/scripts/updatestatistics.py:
```diff
@@ -126,7 +125,7 @@ def get_trending_pages(service, profile_id):
     template_page = pywikibot.Page(pywikibot.Link(template), pywikibot.Site())
     text = " <noinclude>This page is automatically generated. Changes will be overwritten, so '''do not modify'''.</noinclude>\n"
-    for p in pages:
+    for p in pages[:REQD_TEMPLATE_LEN]:
```
I was a bit surprised by this truncation behaviour in the `update_...` function. I think it would be slightly clearer to do the truncation outside this function (ie. in the `main` method), so the only concern in here is actually updating the page.

icyflame replied:
Ah, yes. I will move this into the main function. I didn't think that doing this in the get_popular_pages function was the right choice because (as in this case) we might need to have a larger set from the DB which we will prune based on our policy. So, I am getting all 50 results from the DB using that function.

amrav commented on jobs/pywikibot/scripts/updatestatistics.py:
```diff
@@ -143,6 +142,21 @@ def update_list_of_pages(template, pages):
+    dict_from_list = OrderedDict.fromkeys(list_to_dedup)
+    for elem in base_list:
+        if elem in dict_from_list:
+            list_to_dedup.remove(elem)
```
I assume you're doing the `OrderedDict` stuff to make the complexity O(m + n) instead of O(mn)? If so, doesn't this line make the complexity O(mn)? Should this be `dict_from_list.remove(elem)`?

More generally, I don't think it's worth doing these kinds of optimisations unless really necessary; they make the code harder to understand, more prone to bugs, and don't actually make it practically faster or better. A simple double loop would suffice IMO.

icyflame replied:
Um, no, actually the complexity is still O(m+n) where m is the size of the `base_list` and n is the size of the list to be deduped `list_to_dedup`

1. Form a dictionary from value to bool (say) of the list to be deduplicated `list_to_dedup` - `O(n)`
1. Take every element in the base_list - `O(m)`
    1. Check if it is there in `list_to_dedup` using the dictionary - `O(1)`
    1. If it exists in that list, then remove it - `O(1)`

I did this because we can't iterate over `list_to_dedup` and remove elements from it at the same time.

This approach is faster than O(mn) and I would prefer to stick with it. I think this is a pretty standard approach? (forming a dictionary and removing duplicates) I will add some documentation and then you can see how you feel about the code.

amrav replied:
Isn't step 2. ii. `O(n)` since you're calling `remove` on a `list`?

icyflame replied:
Ah, jeez, sorry sorry! I totally missed that. I looked at my code a few times to see why you thought it was O(mn), I completely missed that. Sorry about that! :cry: 

I have changed the code now. I start with an empty list and add elements that aren't in `base_list` to the new list and return the new list.

amrav reviewed (changes requested):
Thanks for doing this! A couple of minor issues, but looks good otherwise.

icyflame committed:
move truncating step to the main func

icyflame committed:
add documentation about the dedup func

icyflame committed:
fix dedup function to make it O(m+n)

the previous version of the function was O(mn) because removing an element from
a list is O(n). And this removal is done at the worst for every element in the
base_list => O(mn)

icyflame commented:
Ping @amrav :sweat_smile:

amrav commented on jobs/pywikibot/scripts/updatestatistics.py:
```diff
@@ -143,6 +142,22 @@ def update_list_of_pages(template, pages):
+    '''
+
+    deduped_list = [ ]
+    base_dict = OrderedDict.fromkeys(base_list)
```
This can be a `set` now? I thought the way you would use this is:

```python
deduped_list = OrderedDict.fromkeys(list_to_dedup)
for elem in base_list:
  deduped_list.remove(elem) # handle missing keys appropriately
return deduped_list.keys()
```

but if we're just checking membership of the base list, we don't need a dict?

icyflame replied:
Yeah, no need for a dict. I didn't know that set membership checking is also O(1). I have used a set here and removed the `OrderedDict` import. :sweat_drops:

amrav reviewed (approved):
🚢 after addressing comment

amrav commented on jobs/pywikibot/scripts/updatestatistics.py:
```diff
@@ -151,10 +166,20 @@ def main():
-    update_list_of_pages('Template:Trending_pages', trending_pages)
+    trending_pages_deduped = deduplicate_lists( \
+                                trending_pages, \
+                                popular_pages \
```
This should be `popular_pages[:REQD_TEMPLATE_LEN]`, otherwise we're potentially removing trending items that aren't going to show in popular anyway.

icyflame replied:
This truncation is done at L171. 

[`popular_pages = popular_pages[:REQD_TEMPLATE_LEN]`](https://github.com/metakgp/metakgp-wiki/blob/65ad097625f636d85961ca187dc253ccb78f3527/jobs/pywikibot/scripts/updatestatistics.py#L171)

amrav replied:
Ah yes, dammit mutability!

icyflame committed:
use set instead of OrderedDict

set also has O(1) membership checking:

https://stackoverflow.com/a/13884213/2080089
